Feat release 4.1

barbieri opened the pull request:
Prepare release 4.1.0 fixing #52 and updating dependencies

barbieri committed:
test: remove dotenv and use OS-allocated ports in tests

Instead of manually specifying ports using env-vars, which could
conflict with other services, let's use `0` and let the operating
system auto allocate port numbers, then we query these and pass to
users.

barbieri committed:
feat: update straightforward dependencies

barbieri committed:
feat: upgrade @apollo/gateway 2.0.4

Since @apollo/gateway 2.0.4 we cannot use plain `Request` instance
anymore, it's referred directly from the `node-fetch`, which is now a
peer dependency.

It should be harmless since the default `didEncounterError()`, the
only thing that really uses that, just ignores the parameter. Also,
the `node-fetch` version is a superset of the basic `Request`, so
things should work.

barbieri committed:
feat: upgrade graphql-upload to 15.0.1

They changed the export to be ESM, so we must use the whole name with
extensions, changing the `module` in tsconfig.json and the import
lines must include the whole path with extension, just need to disable
eslint import/extensions for such lines.

The typescript is declared as JSDoc, thus tsconfig.json was also
changed to include `maxNodeModuleJsDepth: 10`, otherwise it wasn't
being found.

Convert graphql-upload to a direct runtime dependency so we're fixed
on the correct version and our users don't have to bother to align
that.

barbieri committed:
feat: bump package version

Upgraded dependencies and fixes to cope with newer @apollo/gateway
2.0.4, the most important fix is related to #52, that requires the
headers to be plain objects instead of `Headers` instances.

https://github.com/profusion/apollo-federation-file-upload/issues/52#issuecomment-1146178030

barbieri committed:
ci: improve github workflow

use newer actions and features, such as the "cache: yarn" for
setup-node.

cabelitos commented on tsconfig.json:
```diff
@@ -2,7 +2,7 @@
     "target": "ES2015",
     "moduleResolution": "node",
-    "module": "CommonJS",
+    "module": "nodenext",
```
you sure about this one? Not sure if this breaks old note versions like 12.

barbieri replied:
I'm not that sure, but comparing the output file it's the same.

As for node12, it's already not supported by Apollo itself, then we're 14 and up.

cabelitos commented on package.json:
```diff
@@ -22,55 +22,54 @@
     "apollo-server-types": "^3.5.2",
-    "graphql-upload": "^13.0.0"
+    "graphql-upload": "^13.0.0",
+    "node-fetch": "^3.2.5"
```
I was wondering. What do you think about this as dep and not a peer dep?

douglas-wang replied:
Quick question, is there any plan to upgrade this to the latest version?
https://www.npmjs.com/package/graphql-upload

barbieri replied:
@cabelitos could be, I kept it as before, but I think it may be worth to have as dep as it's only use directly (unlike `@apollo/gateway` that is use by our users).

@douglas-wang it's already 15.0.1 at the devDependency, but I'll move this to a simple direct dependency.

cabelitos commented on package.json:
```diff
@@ -22,55 +22,54 @@
     "apollo-server-env": "^4.2.1",
-    "apollo-server-express": "^3.6.7",
-    "apollo-server-types": "^3.5.2",
-    "dotenv-safe": "^8.2.0",
```
remember to delete the .env.example

cabelitos replied:
Ignore me. it was deleted. lol.

bragovo commented:
Hello! When you planning to release this? We have a lot of pain about https://github.com/profusion/apollo-federation-file-upload/issues/52 :-D

barbieri committed:
fix: simplify types after @apollo/gateway-2.0.4

This is base on conversations:
 - https://github.com/profusion/apollo-federation-file-upload/issues/52#issuecomment-1148946002
 - https://github.com/profusion/apollo-federation-file-upload/issues/52#issuecomment-1148973969

It removes some dependencies, namely node-fetch as those classes are
not used and just add overhead.

It also removes the option to specify `fetcher` and uses
RemoteGraphQLDataSource default (make-fetch-happen)

barbieri commented:
> Hello! When you planning to release this? We have a lot of pain about #52 :-D

hi @bragovo, just sent a new try based on https://github.com/profusion/apollo-federation-file-upload/issues/52#issuecomment-1148973969, I'm waiting for @glasser to review so I can merge and release.

burn2delete committed:
fix: send custom headers added in `willSendRequest`

- This commit makes so that headers added in `willSendRequest` are not
ignored if the request had no headers

barbieri commented:
@glasser, @bragovo would you mind confirming this works? I'll merge the fix proposed in #57 and release soon (today or tomorrow)

bragovo commented:
@barbieri unfortunately my JS teammates still can't upload anything :/ I need some more time to check how to fix it.

barbieri commented:
> @barbieri unfortunately my JS teammates still can't upload anything :/ I need some more time to check how to fix it.

weird, wasn't the `Headers` the problem? It should be using the plain `Object` now. Just check if the #57 pointed issue is not the problem (in other words: if you're relying on headers set at `willSendRequest`, we need that fix so it works.

rostber commented:
That pull make problem. The `willSendRequest` method will call with `undefined` headers object (in original RemoteGraphQLDataSource headers always belongs to Headers and never `undefined`).

By the way, FileUploadDataSource class doesn't send `operationName`. Way to fix it:
```
    const operations = JSON.stringify({
      query: request.query,
      variables,
      operationName:
        'incomingRequestContext' in args
          ? args.incomingRequestContext.request.operationName
          : args.request.operationName,
    });
    ```

TheBit commented:
Hi, any news on this?
We can't go live without upload :(

sfounder commented:
Hello. Is it so critical to merge? I see few developers are blocked this MR. We also can't go live without upload. Could you merge this MR and then refactor code? Thanks!

paulpdaniels commented:
Any help that the maintainers need here? I know we all get busy, so anything I can do to help move this along?

glasser commented:
Hi, I found a bit of time to look at this but I'm not sure what I'm being asked to review. It looks like you're working around the `Fetcher` type declaration with an `as` typecast, which you need because you're trying to put a FormData as the body which `Fetcher` doesn't expect. But [as I said](https://github.com/profusion/apollo-federation-file-upload/issues/52#issuecomment-1148973969) I don't think that arbitrary Fetch implementations all handle FormData in the same way. So doing it this way is essentially locking your API in to only working with Fetch implementations that happen to work with the FormData implementation you chose. To me this seems likely to lead to confusion. My [recommendation continues to be](https://github.com/profusion/apollo-federation-file-upload/issues/52#issuecomment-1148973969) that if this package depends on the ability to send FormData (as implemented by the `form-data` npm package) via a Fetch API, then it should just choose a particular implementation of Fetch that definitely supports the `form-data` package and use it directly instead of allowing the fetcher to be overridden with an arbitrary implementation that might not work.

glasser commented:
(That said, if you're OK with the idea that your package will be fragile and could break if a user passes in a Fetcher implementation that doesn't support FormData from `form-data` as a body, then this is probably fine? I'm not an expert on the details of your project but if it works and passes tests then it's probably OK?)

DDDKnightmare commented on lib/FileUploadDataSource.ts:
```diff
@@ -1,13 +1,15 @@
 import { isObject } from '@apollo/gateway/dist/utilities/predicates';
 import cloneDeep from 'lodash.clonedeep';
 import set from 'lodash.set';
+import type { FetchInterface, FetchOptions } from 'make-fetch-happen';
```
Considering what has been suggested [here](https://github.com/profusion/apollo-federation-file-upload/pull/55#issuecomment-1262791533), wouldn't it be better to add `make-fetch-happen` as direct dependency and force it to be used instead of using the default from apollo? (it uses `make-fetch-happen` at the moment, but this can change in future versions).

TheBit commented:
@cabelitos sorry for disturbing, but this is waiting for half a year, are there any plans for this?

BobbiSixkiller commented:
Is this going to be released any time soon please ?

oliveirarleo commented:
Hello @TheBit and @BobbiSixkiller. We cannot expect help from @cabelitos anymore, but our team at ProFUSION will put some hands here to help you guys out right at our next sprint. We now have officially including this task in our planning. That means we'll start working (and revise the work done here) as of 30th of January and we expect to solve the problem and merge between 2 and 4 weeks.

g7fernandes commented on lib/FileUploadDataSource.ts:
```diff
@@ -196,34 +202,45 @@ export default class FileUploadDataSource extends RemoteGraphQLDataSource {
     await this.addDataHandler(form, resolvedFiles);
 
-    const headers = (request.http && request.http.headers) || new Headers();
+    // This must happen before constructing the request headers
+    // otherwise any custom headers set in willSendRequest are ignored
+    if (this.willSendRequest) {
+      await this.willSendRequest(args);
+    }
```
test covering this here https://github.com/profusion/apollo-federation-file-upload/pull/67

oliveirarleo commented:
:tada:
